File: src/main/java/ru/arc/hooks/luckperms/LuckPermsHook.java
package ru.arc.hooks.luckperms;

import lombok.extern.slf4j.Slf4j;
import net.luckperms.api.LuckPermsProvider;
import net.luckperms.api.model.group.Group;
import net.luckperms.api.model.user.UserManager;
import net.luckperms.api.node.types.MetaNode;
import net.luckperms.api.query.Flag;
import net.luckperms.api.query.QueryMode;
import net.luckperms.api.query.QueryOptions;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;

import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

import static ru.arc.util.Logging.error;

@Slf4j
public class LuckPermsHook {

    public boolean hasPermission(OfflinePlayer offlinePlayer, String perm) {
        UserManager userManager = LuckPermsProvider.get().getUserManager();
        if (offlinePlayer instanceof Player player) {
            return player.hasPermission(perm);
        }
        try {
            if (Thread.currentThread().getName().contains("main")) {
                System.out.println("Loading permission data from main thread!!!");
            }
            return userManager.loadUser(offlinePlayer.getUniqueId()).get()
                    .getCachedData()
                    .getPermissionData()
                    .checkPermission(perm)
                    .asBoolean();
        } catch (InterruptedException | ExecutionException e) {
            error("Error while checking permission", e);
            return false;
        }
    }

    public List<String> getGroups(OfflinePlayer offlinePlayer) {
        UserManager userManager = LuckPermsProvider.get().getUserManager();
        try {
            if (Thread.currentThread().getName().contains("main")) {
                error("Loading groups data from main thread!!!");
            }
            return userManager.loadUser(offlinePlayer.getUniqueId()).get()
                    .getInheritedGroups(QueryOptions.builder(QueryMode.NON_CONTEXTUAL)
                            .flag(Flag.RESOLVE_INHERITANCE, true)
                            .build()).stream()
                    .map(Group::getName)
                    .toList();
        } catch (InterruptedException | ExecutionException e) {
            error("Error while getting groups", e);
            return List.of();
        }
    }

    public CompletableFuture<Void> setMeta(UUID uuid, String key, String value) {
        UserManager userManager = LuckPermsProvider.get().getUserManager();
        return userManager.modifyUser(uuid, user -> {
            user.getNodes().stream()
                    .filter(node -> node instanceof MetaNode)
                    .map(node -> (MetaNode) node)
                    .filter(node -> node.getMetaKey().equals(key))
                    .forEach(node -> user.data().remove(node));
            if (value == null) return;
            MetaNode node = MetaNode.builder()
                    .key(key)
                    .value(value)
                    .build();
            user.data().add(node);
        });
    }

    public CompletableFuture<String> getMeta(UUID uuid, String key) {
        UserManager userManager = LuckPermsProvider.get().getUserManager();
        return userManager.loadUser(uuid)
                .thenApply(user -> user.getCachedData().getMetaData().getMetaValue(key));
    }

}

File: src/main/kotlin/ru/arc/hooks/luckperms/LuckPermsHook.kt
package ru.arc.hooks.luckperms

import net.luckperms.api.LuckPermsProvider
import net.luckperms.api.model.group.Group
import net.luckperms.api.model.user.User
import net.luckperms.api.model.user.UserManager
import net.luckperms.api.node.types.MetaNode
import net.luckperms.api.query.Flag
import net.luckperms.api.query.QueryMode
import net.luckperms.api.query.QueryOptions
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import ru.arc.util.Logging
import ru.arc.util.Logging.error
import ru.arc.util.Logging.warn
import java.util.UUID
import java.util.concurrent.CompletableFuture

class LuckPermsHook(
    private val userManager: UserManager = LuckPermsProvider.get().userManager,
    private val isPrimaryThread: () -> Boolean = Bukkit::isPrimaryThread,
    private val queryOptions: () -> QueryOptions = {
        QueryOptions
            .builder(QueryMode.NON_CONTEXTUAL)
            .flag(Flag.RESOLVE_INHERITANCE, true)
            .build()
    },
) {
    fun hasPermission(
        offlinePlayer: OfflinePlayer,
        perm: String,
    ): Boolean {
        if (offlinePlayer is Player) return offlinePlayer.hasPermission(perm)
        return try {
            if (Bukkit.isPrimaryThread()) {
                warn("Loading permission data from main thread!")
            }
            userManager
                .loadUser(offlinePlayer.uniqueId)
                .get()
                .cachedData.permissionData
                .checkPermission(perm)
                .asBoolean()
        } catch (e: Exception) {
            error("Error while checking permission", e)
            false
        }
    }

    fun getGroups(offlinePlayer: OfflinePlayer): List<String> {
        return try {
            resolveUserForRead(offlinePlayer)?.getInheritedGroups(
                queryOptions(),
            )?.map(Group::getName).orEmpty()
        } catch (e: Exception) {
            error("Error while getting groups", e)
            emptyList()
        }
    }

    private fun resolveUserForRead(offlinePlayer: OfflinePlayer): User? {
        userManager.getUser(offlinePlayer.uniqueId)?.let { return it }
        if (isPrimaryThread()) return null
        return userManager.loadUser(offlinePlayer.uniqueId).get()
    }

    fun setMeta(
        uuid: UUID,
        key: String,
        value: String?,
    ): CompletableFuture<Void> {
        val userManager = LuckPermsProvider.get().userManager
        return userManager.modifyUser(uuid) { user ->
            user.nodes
                .filterIsInstance<MetaNode>()
                .filter { it.metaKey == key }
                .forEach { user.data().remove(it) }
            if (value == null) return@modifyUser
            val node =
                MetaNode
                    .builder()
                    .key(key)
                    .value(value)
                    .build()
            user.data().add(node)
        }
    }

    fun getMeta(
        uuid: UUID,
        key: String,
    ): CompletableFuture<String?> {
        val userManager = LuckPermsProvider.get().userManager
        return userManager
            .loadUser(uuid)
            .thenApply { it.cachedData.metaData.getMetaValue(key) }
    }
}
